NEW: Expose silverstripe-backend for modules to access.

sminnee opened the pull request:
FIX: Ensure post has correct content-type header.

Exporting silverstripe-backend lets other modules (such as asset-admin)
use it.

In addition, the Content-type header of a post request of URL-encoded
data was being set to text/plain by default, which isn’t correct and
stopped PHP from interpreting it.

clarkepaul committed:
initial updates to main nav styles for 4.0

Conflicts:
	admin/css/ie7.css
	admin/css/ie7.css.map
	admin/css/ie8.css.map
	admin/css/screen.css
	admin/css/screen.css.map
	admin/javascript/dist/bundle-leftandmain.js
	admin/javascript/src/LeftAndMain.Menu.js
	admin/scss/_spritey.scss
	css/GridField.css

Commit:
Include bootstrap throughout the CMS

Commit:
Fix ability to edit files immeditately after upload

Conflicts:
	javascript/dist/UploadField.js

davidmorrisonnz committed:
Tidy grunt file

- Don't watch in prod mode
- Use gulp-if instead of if statements

davidmorrisonnz committed:
Adds schemaDataType to FormFieldSchema

davidmorrisonnz committed:
Fix bottom toolbar height

davidmorrisonnz committed:
CampaignAdmin and GridField React sections

Also removes watchify because it wasn't working.
Add SilverStripeBackend class used to fetch data from endpoints for the front-end

Commit:
Fix menu dropdown icon

Commit:
GridField and header styles

chillu committed:
Fix form schema ID getter

Also moved keys of higher importance to start of array,
easier to review this way

chillu committed:
Only iterate through dataFields() in FormSchema

Temporary measure until we implement nested fields

chillu committed:
Defined $schemaDataType constant, added to FormField subclasses

clarkepaul committed:
Add generic React components

Includes moving some components from AssetAdmin

davidmorrisonnz committed:
Client config

chillu committed:
Campaign form schema and edit form

Remove once we have a schema-driven GridField in place.
Edit form should be a GridFieldDetailForm, but there's more work to do for schema generation
from this component.

chillu committed:
Default to schema response part in LeftAndMain->schema()

More of a standard API approach to return data by default, and make customisation
via HTTP headers an optional mode.

chillu committed:
Simplified form schema structure and mocks

Keyed by URL instead of anonymous object maps which need to be iterated on
Removed the 'schema.forms' namespace, unnecessary since all 'schema' items should be forms

davidmorrisonnz committed:
Consolidated bundles, sourcemaps and watching

We don't want a bundle for each section, unnecessary network overhead
and too much boilerplate in defining build tooling.

Commit:
- Add GridfieldActionsComponent
- Add ability to set desired relative width on gridfield cells

chillu committed:
Consolidated grid-field component

Every top-level folder in /components should be a standalone component.
This isn't the case for the grid-field-* ones, so they belong into the same folder.
See https://github.com/Automattic/wp-calypso/blob/master/docs/components.md

merge styles

chillu committed:
Relative ES6 module paths

http://stackoverflow.com/questions/20158401/how-do-i-manage-relative-path-aliasing-in-multiple-grunt-browserify-bundles/23608416#23608416
https://github.com/substack/node-browserify/issues/767
http://stackoverflow.com/questions/28078780/relative-paths-in-package-json
https://github.com/npm/npm/issues/1558
http://stackoverflow.com/questions/31624567/webstorm-es6-named-import-getting-cannot-resolve-symbol-error/31631415
http://stackoverflow.com/questions/20158401/how-do-i-manage-relative-path-aliasing-in-multiple-grunt-browserify-bundles/23608416#23608416
https://github.com/vigetlabs/gulp-starter/issues/17#issuecomment-44292681

lozcalver committed:
API: Remove deprecated caching behaviour from ViewableData (closes #4063)

chillu committed:
Watch admin SCSS files

chillu committed:
Removed duplicate bootstrap lib

Now generated on the fly through framework/admin/scss

chillu committed:
Removed sourcemaps

Will be part of .gitignore in a subsequent commit

chillu committed:
Moved FormFieldSchemaTrait into FormField

The RFC requires a FormField implementation to override $schemaDataType,
but its defined on the trait - which can't be redefined by a field subclass.

In the end, the trait was never designed to be reuseable on classes other than FormField.
We need to admit that architecturally, we'll have to add all that API weight to the base FormField
class because of the way forms are structured in SilverStripe (mainly due to a missing layer
of indirection in getCMSFields implementations).

Also implemented the $schemaDataType on fields where its known.

See https://github.com/silverstripe/silverstripe-framework/issues/4938
See http://php.net/manual/en/language.oop5.traits.php#language.oop5.traits.properties.example

davidmorrisonnz committed:
Implement client-side routing with Page.js

- Removes thirdparty dependency History.js
- Adds thirdparty dependency Page.js to manage client-side routing
- Adds a wrapper around Page.js for SilverStripe specific behaviour
- Increased minimum browser requirement to IE10. Native HTML History API routing requires IE10 or newer (necessitated by removal of History.js)
- PJAX pannel loading via now uses promises rather than callbacks
- Adds getClientConfig method to LeftAndMain which can be used to pass config from to the front-end client

chillu committed:
Generic state management for React GridField

Renaming state operations from 'campaign' to 'record'.
Implemented API endpoint retrieval of GridField data.

Added more mock data into CampaignAdmin (rather than hardcoding in client),
to be replaced by CampaignAdmin API endpoint querying the real datamodel.

Using more native isomorphic-fetch instead of jQuery.ajax
to minimise dependencies and get into a more forward-thinking API.

Also catching errors in ReactJS API backend:
Emulate jQuery.ajax() behaviour. Might change at a later point
if we have a general purpose backend with a promise-based catch()
implementation.

Commit:
Implement readCampaigns()
Renamed "item/items" urls to "set/sets" to distinguish between change sets and change set items.

Commit:
Move ActionComponent into FormActionComponent

Commit:
Merge pull request #5238 from open-sausages/pulls/4.0/build-tooling-cleanup

Build tooling cleanup

Commit:
API CMSMenu::get_cms_classes() is now sorted

Commit:
Merge pull request #5237 from open-sausages/pulls/4.0/bootstrap-include-default

Include Bootstrap by default

chillu committed:
Use /index suffix for JS import statements

Otherwise browserify or gulp-sourcemaps doesn't include these
in the *.map file, which makes debugging impossible

chillu committed:
Delete campaign feature

chillu committed:
Updated react dependency

chillu committed:
Updated babel dependency

chillu committed:
More logical dependency ordering in Gulpfile

Commit:
Merge pull request #5239 from open-sausages/pulls/4.0/form-schema-fixes

Form schema fixes

Commit:
Merge pull request #5232 from kinglozzer/viewabledata-dep

Remove deprecated caching behaviour from ViewableData (closes #4063)

Commit:
API Add $context method to DataObject::can for consistency with canCreate()

Commit:
BUG Fix issue with SapphireTest::assertDOSEquals incorrectly passing on empty set

Commit:
API SapphireTest::logInWithPermission now supports multiple permissions

Commit:
API Implement ChangeSets for batch publishing

chillu committed:
Add ESLint support

See https://github.com/silverstripe/silverstripe-framework/pull/5108

Commit:
BUG Prevent live manifest loading files from nested test directories
E.g. framework/admin/tests

chillu committed:
Merge pull request #5254 from tractorcow/pulls/4.0/ignore-tests

BUG Prevent live manifest loading files from nested test directories

Commit:
Convert react grid-field from flexbox to table-based

Commit:
BUG move test file to correct folder and fix class_exists

chillu committed:
Merge pull request #5255 from tractorcow/pulls/4.0/fix-test-location

BUG move test file to correct folder and fix class_exists

steve-silverstripe committed:
Changed CMS close from an 'x' to a silverstripe font icon

Commit:
Merge pull request #5257 from steve-silverstripe/update-close-icon-on-cms-notices

Changed CMS close from an 'x' to a silverstripe font icon

chillu committed:
Merge pull request #5243 from open-sausages/pulls/4.0/client-routing

Implement client-side routing with Page.js

Commit:
Updates to FormActionComponent

Commit:
Moving thirdparty module to npm dependency - json-js.

Commit:
API Update Versioned methods

Commit:
TinyMCE 4 optimisations. Use TinyMCE_Compressor.

Commit:
NEW: Expose silverstripe-backend for modules to access.
FIX: Ensure post has correct content-type header.

Exporting silverstripe-backend lets other modules (such as asset-admin)
use it.

In addition, the Content-type header of a post request of URL-encoded
data was being set to text/plain by default, which isn’t correct and
stopped PHP from interpreting it.

chillu committed:
Reapplied customisations to tiny_mce_gzip.php

See previously applied customisations at https://github.com/silverstripe/silverstripe-framework/blob/3/thirdparty/tinymce/tiny_mce_gzip.php
See https://github.com/silverstripe/silverstripe-framework/pull/5258

chillu committed:
Partially reverted b1b85b5

We still require copying thirdparty components from node_modules on tiny_mce

chillu committed:
Reinstate route externals in Gulpfile

Accidental regression from c4e5074f67 (see aa1aebd0079)

Commit:
Merge pull request #5063 from benmanu/pulls/add-npm-json-js

Moving thirdparty module to npm dependency - json-js

chillu committed:
Fix IE compat through babelify

Requires a polyfill for Object.assign

Commit:
Don't use removed API in install.php

Commit:
Merge pull request #5247 from open-sausages/pulls/4.0/changesets

API Implement ChangeSets for batch publishing

lozcalver committed:
Merge pull request #5263 from tractorcow/pulls/4.0/install-sync

Don't use removed API in install.php

chillu committed:
Always use uglify during build

It makes very little difference between "npm run build"
and "npm run build --development" (both under a second),
since the gulp pipeline is smart enough to only uglify the new bits.

Creating different dist files with "--development"
is causing grief during pull requests, since most
devs will add the changed files without reviewing them.
It also means you can commit without stopping your "watch" npm task.

flashbackzoo commented on admin/javascript/src/silverstripe-backend.js:
```diff
@@ -45,8 +45,15 @@ class SilverStripeBackend {
-      .then(checkStatus);
+    return this.fetch(url, {
+      method: 'post',
+      headers: new Headers({
```
I can't see `Headers` defined anywhere?

sminnee replied:
@flashbackzoo from the looks of things, Headers is added to the globals when `isomorphic-fetch` is included. :-/ https://github.com/matthew-andrews/isomorphic-fetch/blob/master/fetch-npm-node.js

chillu commented on admin/javascript/src/silverstripe-backend.js:
```diff
@@ -45,8 +45,15 @@ class SilverStripeBackend {
+    return this.fetch(url, {
+      method: 'post',
+      headers: new Headers({
+        'Content-Type': 'application/x-www-form-urlencoded',
```
Hm, I think hardcoding a header here isn't ideal. How about `Object.assign({}, headers, {'Content-Type': 'application/x-www-form-urlencoded'})` with `headers` passed in as an optional method argument? That should also apply to the other HTTP verbs in the backend here.

Commit:
BUG Fix error when using search bar

chillu committed:
Fix non-standard url encoding in CMS search form

jQuery uses encodeURIComponent, but also this line here:
return s.join( "&" ).replace( r20, "+" );
https://github.com/jquery/jquery/blob/1.7.2/src/ajax.js#L797

This breaks when passing the result through $.path.addSearchParams(),
which in turn uses jQuery.param - any '%20' would've been wrongly encoded as '+' by jQuery.serialise(),
which in turn gets rightly encoded as '%2B' by jQuery.param().

This had the effect of breaking CMS searches with spaces in them.

chillu committed:
Include config in bundles

Merge regression

chillu committed:
Don't show diff for compiled CSS files

chillu committed:
Added TODO elements to unimplemented CampaignAdmin functionality

chillu committed:
Enforcing can*() in CampaignAdmin reads

chillu committed:
Syntax fix

chillu committed:
Correct state handling in <FormBuilder> component

chillu committed:
More accurate variable naming

schema.schema is awkward

chillu committed:
Only list existing <FormBuilder> data types

chillu committed:
Allow user-defined <Form> attributes

chillu committed:
Pass all field attributes via <FormBuilder>

Also make them immutable via deep-freeze

chillu committed:
More concise breadcrumb generation

Courtesy of @hafriedlander

chillu committed:
Removed unnecessary thunk wrapper in actions

chillu committed:
Updated build artifacts
